File: swift/Sources/OpenID4VP/HeldSdJwtVc.swift
import Crypto
import Foundation
import SdJwt

/// An SD-JWT VC the wallet holds, usable as a DCQL match target and presentable over OpenID4VP.
public struct HeldSdJwtVc: QueryableCredential {
    public let credentialId: String
    let sdJwt: SdJwt
    let holderSigner: any JwsSigner

    public let format = "dc+sd-jwt"
    public let claims: JsonValue
    public let docType: String? = nil

    public var vct: String? {
        if case let .str(v)? = claims["vct"] { return v }
        return nil
    }

    public init(credentialId: String, sdJwt: SdJwt, holderSigner: any JwsSigner) throws {
        self.credentialId = credentialId
        self.sdJwt = sdJwt
        self.holderSigner = holderSigner
        self.claims = try SdJwtHolder.processedClaims(sdJwt)
    }

    /// Selects the `disclosedPaths` disclosures and appends a KB-JWT bound to `audience` + `nonce`.
    public func present(
        disclosedPaths: [[String]],
        audience: String,
        nonce: String,
        issuedAt: Int64,
        transactionData: [String]? = nil
    ) async throws -> String {
        let pathSet = Set(disclosedPaths)
        var extra: [(String, JsonValue)] = []
        if let td = transactionData, !td.isEmpty {
            extra.append(("transaction_data_hashes", .arr(td.map { .str(sha256B64($0)) })))
            extra.append(("transaction_data_hashes_alg", .str("sha-256")))
        }
        let presented = try await SdJwtHolder.presentWithKeyBinding(
            sdJwt, select: { pathSet.contains($0) },
            audience: audience, nonce: nonce, issuedAt: issuedAt, signer: holderSigner, extraClaims: extra
        )
        return presented.serialize()
    }

    private func sha256B64(_ s: String) -> String {
        Base64Url.encode([UInt8](SHA256.hash(data: Data(s.utf8))))
    }
}

File: kotlin/openid4vp/src/main/kotlin/com/hopae/eudi/wallet/vp/HeldSdJwtVc.kt
package com.hopae.eudi.wallet.vp

import com.hopae.eudi.wallet.sdjwt.Base64Url
import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.sdjwt.JwsSigner
import com.hopae.eudi.wallet.sdjwt.SdJwt
import com.hopae.eudi.wallet.sdjwt.SdJwtHolder
import java.security.MessageDigest

/** An SD-JWT VC the wallet holds, usable as a DCQL match target and presentable over OpenID4VP. */
class HeldSdJwtVc(
    override val credentialId: String,
    val sdJwt: SdJwt,
    private val holderSigner: JwsSigner,
) : PresentableCredential {

    override val format: String = "dc+sd-jwt"
    override val claims: JsonValue.Obj = SdJwtHolder.processedClaims(sdJwt)
    override val vct: String? = (claims["vct"] as? JsonValue.Str)?.value
    override val docType: String? = null

    /**
     * Builds a presentation: selects the disclosed disclosures and appends a KB-JWT bound to
     * the verifier client_id + nonce, optionally with transaction-data hashes.
     */
    override suspend fun present(ctx: PresentationContext): String {
        val pathSet = ctx.disclosedPaths.toSet()
        val extra = buildList {
            if (!ctx.transactionData.isNullOrEmpty()) {
                add("transaction_data_hashes" to JsonValue.Arr(ctx.transactionData.map { JsonValue.Str(sha256B64(it)) }))
                add("transaction_data_hashes_alg" to JsonValue.Str("sha-256"))
            }
        }
        val presented = SdJwtHolder.presentWithKeyBinding(
            issued = sdJwt,
            select = { it in pathSet },
            audience = ctx.clientId,
            nonce = ctx.nonce,
            issuedAt = ctx.issuedAt,
            signer = holderSigner,
            extraClaims = extra,
        )
        return presented.serialize()
    }

    private fun sha256B64(s: String): String =
        Base64Url.encode(MessageDigest.getInstance("SHA-256").digest(s.encodeToByteArray()))
}
